File: smartadapter/src/main/java/smartadapter/SmartAdapterBuilder.java
package smartadapter;

/*
 * Created by Manne Öhlund on 29/05/17.
 * Copyright © 2019 All rights reserved.
 */

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.HashMap;

import smartadapter.internal.factory.SmartRecyclerAdapterExtensionFactory;
import smartadapter.internal.mapper.ViewEventMapper;
import smartadapter.listener.OnItemSelectedListener;
import smartadapter.listener.OnViewEventListener;
import smartadapter.viewholder.SmartViewHolder;
import smartadapter.widget.ViewTypeResolver;

/**
 * Builder for SmartRecyclerAdapter.
 */
public class SmartAdapterBuilder {

    private SmartRecyclerAdapter smartRecyclerAdapter;
    private RecyclerView.LayoutManager layoutManager;
    private ViewTypeResolver viewTypeResolver;
    private HashMap<String, Class<? extends SmartViewHolder>> viewHolderMapper = new HashMap<>();
    private HashMap<Class<? extends SmartViewHolder>, SmartRecyclerAdapter> smartRecyclerAdapterMapper = new HashMap<>();
    private ViewEventMapper viewEventMapper = new ViewEventMapper();
    private SmartRecyclerAdapterExtensionFactory smartRecyclerAdapterExtensionFactory = new SmartRecyclerAdapterExtensionFactory();

    SmartAdapterBuilder(@NonNull SmartRecyclerAdapter smartRecyclerAdapter) {
        this.smartRecyclerAdapter = smartRecyclerAdapter;
    }

    public final SmartAdapterBuilder map(@NonNull Class<?> itemType, @NonNull Class<? extends SmartViewHolder> viewHolderType) {
        viewHolderMapper.put(itemType.getName(), viewHolderType);
        return this;
    }

    public final SmartAdapterBuilder map(@NonNull Class<? extends SmartViewHolder> viewHolderType, @NonNull SmartRecyclerAdapter smartRecyclerAdapter) {
        smartRecyclerAdapterMapper.put(viewHolderType, smartRecyclerAdapter);
        return this;
    }

    public final SmartAdapterBuilder setLayoutManager(@NonNull RecyclerView.LayoutManager layoutManager) {
        this.layoutManager = layoutManager;
        return this;
    }

    private RecyclerView.LayoutManager getLayoutManager(Context context) {
        if (layoutManager == null) {
            layoutManager = new LinearLayoutManager(context);
        }
        return layoutManager;
    }

    public final SmartAdapterBuilder setViewTypeResolver(@NonNull ViewTypeResolver viewTypeResolver) {
        this.viewTypeResolver = viewTypeResolver;
        return this;
    }

    /**
     * Adds {@link OnViewEventListener} to the {@link SmartRecyclerAdapter}.
     * The adapter will then automatically map the {@link OnViewEventListener} to the target view holder class with {@link OnViewEventListener#getViewHolderType()},
     * set the viewActionListener on the right View with viewId using {@link OnViewEventListener#getViewId()}.
     *
     * @see OnViewEventListener#getViewEventId() for predefined eventIds.
     *
     * @param viewEventListener target {@link OnViewEventListener}
     * @return SmartAdapterBuilder
     */
    public final SmartAdapterBuilder addViewEventListener(@NonNull OnViewEventListener viewEventListener) {
        if (viewEventListener instanceof OnItemSelectedListener) {
            ((OnItemSelectedListener)viewEventListener).getSelectionStateHolder().setSmartRecyclerAdapter(smartRecyclerAdapter);
            viewEventMapper.addViewEventListener(viewEventListener);
        } else {
            viewEventMapper.addViewEventListener(viewEventListener);
        }
        return this;
    }

    /**
     * Adds {@link SmartRecyclerAdapterExtensionBuilder} to {@link SmartRecyclerAdapterExtensionFactory} that will build and bind
     * the {@link SmartRecyclerAdapter} and {@link RecyclerView} to the ExtensionBuilder.
     *
     * @param smartExtensionBuilder extension builder
     * @return SmartAdapterBuilder
     */
    public final SmartAdapterBuilder addExtensionBuilder(@NonNull SmartRecyclerAdapterExtensionBuilder smartExtensionBuilder) {
        smartRecyclerAdapterExtensionFactory.add(smartExtensionBuilder);
        return this;
    }

    @SuppressWarnings("unchecked")
    public final <T> T into(RecyclerView recyclerView) {
        smartRecyclerAdapter.setDataTypeViewHolderMapper(viewHolderMapper);
        smartRecyclerAdapter.setSmartRecyclerAdapterMapper(smartRecyclerAdapterMapper);
        smartRecyclerAdapter.setViewTypeResolver(viewTypeResolver);
        smartRecyclerAdapter.setViewEventMapper(viewEventMapper);
        recyclerView.setAdapter(smartRecyclerAdapter);
        recyclerView.setLayoutManager(getLayoutManager(recyclerView.getContext()));
        smartRecyclerAdapterExtensionFactory.build(smartRecyclerAdapter, recyclerView);
        return (T) smartRecyclerAdapter;
    }

    @SuppressWarnings("unchecked")
    public final <T> T create() {
        smartRecyclerAdapter.setDataTypeViewHolderMapper(viewHolderMapper);
        smartRecyclerAdapter.setSmartRecyclerAdapterMapper(smartRecyclerAdapterMapper);
        smartRecyclerAdapter.setViewTypeResolver(viewTypeResolver);
        smartRecyclerAdapter.setViewEventMapper(viewEventMapper);
        return (T) smartRecyclerAdapter;
    }
}

File: smartadapter/src/main/kotlin/smartadapter/SmartAdapterBuilder.kt
package smartadapter

/*
 * Created by Manne Öhlund on 29/05/17.
 * Copyright © 2019 All rights reserved.
 */

import android.content.Context
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import smartadapter.binders.ItemTouchBinder
import smartadapter.binders.SmartRecyclerAdapterExtension
import smartadapter.internal.extension.isMutable
import smartadapter.internal.extension.name
import smartadapter.viewholder.SmartAdapterHolder
import smartadapter.widget.ViewTypeResolver
import java.util.HashMap
import kotlin.reflect.KClass

/**
 * Builder for [SmartRecyclerAdapter].
 */
open class SmartAdapterBuilder {
    internal var items: MutableList<Any> = mutableListOf()
    internal var layoutManager: RecyclerView.LayoutManager? = null
    internal var viewTypeResolver: ViewTypeResolver? = null
    internal val viewHolderMapper = HashMap<String, SmartViewHolderType>()
    internal val viewHolderBinders = mutableListOf<SmartViewHolderBinder>()
    internal val smartRecyclerAdapterExtensions = mutableMapOf<Any, SmartRecyclerAdapterExtension>()

    internal open fun getSmartRecyclerAdapter() = SmartRecyclerAdapter(items.let {
        (if (it.isMutable()) it else it.toMutableList())
    })

    fun map(itemType: KClass<*>, viewHolderType: SmartViewHolderType): SmartAdapterBuilder {
        viewHolderMapper[itemType.name] = viewHolderType
        return this
    }

    fun setItems(items: List<Any>): SmartAdapterBuilder {
        this.items = (if (items.isMutable()) items else items.toMutableList()) as MutableList<Any>
        return this
    }

    fun setLayoutManager(layoutManager: RecyclerView.LayoutManager): SmartAdapterBuilder {
        this.layoutManager = layoutManager
        return this
    }

    private fun getLayoutManager(context: Context): RecyclerView.LayoutManager {
        if (layoutManager == null) {
            layoutManager = LinearLayoutManager(context)
        }
        return layoutManager!!
    }

    fun setViewTypeResolver(viewTypeResolver: ViewTypeResolver): SmartAdapterBuilder {
        this.viewTypeResolver = viewTypeResolver
        return this
    }

    /**
     * Adds [SmartViewHolderBinder] to the adapter.
     *
     * @param viewHolderBinder extension view holder binder
     * @return SmartAdapterBuilder
     */
    fun addBinder(viewHolderBinder: SmartViewHolderBinder): SmartAdapterBuilder {
        viewHolderBinders.add(viewHolderBinder)
        return this
    }

    /**
     * Adds [SmartRecyclerAdapterExtension] to the adapter with [SmartRecyclerAdapterExtension.identifier] as key.
     *
     * @param extension extension
     * @return SmartAdapterBuilder
     */
    fun addExtension(extension: SmartRecyclerAdapterExtension): SmartAdapterBuilder {
        smartRecyclerAdapterExtensions[extension.identifier] = extension
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> into(recyclerView: RecyclerView): T {
        val smartRecyclerAdapter = getSmartRecyclerAdapter()
        smartRecyclerAdapter.setDataTypeViewHolderMapper(viewHolderMapper)
        smartRecyclerAdapter.viewTypeResolver = viewTypeResolver
        recyclerView.adapter = smartRecyclerAdapter
        recyclerView.layoutManager = getLayoutManager(recyclerView.context)
        viewHolderBinders.forEach {
            smartRecyclerAdapter.addBinder(it)
            (it as? SmartAdapterHolder)?.smartRecyclerAdapter = smartRecyclerAdapter
            (it as? ItemTouchBinder<*>)?.bind(smartRecyclerAdapter, recyclerView)
        }
        smartRecyclerAdapterExtensions.forEach {
            smartRecyclerAdapter.addExtension(it.value)
        }
        return smartRecyclerAdapter as T
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> create(): T {
        val smartRecyclerAdapter = getSmartRecyclerAdapter()
        smartRecyclerAdapter.setDataTypeViewHolderMapper(viewHolderMapper)
        smartRecyclerAdapter.viewTypeResolver = viewTypeResolver
        viewHolderBinders.forEach {
            smartRecyclerAdapter.addBinder(it)
            (it as? SmartAdapterHolder)?.smartRecyclerAdapter = smartRecyclerAdapter
        }
        smartRecyclerAdapterExtensions.forEach {
            smartRecyclerAdapter.addExtension(it.value)
        }
        return smartRecyclerAdapter as T
    }
}
